pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/SpoolIntegrationTest.java

```diff
@@ -0,0 +1,143 @@
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+
+public class SpoolIntegrationTest extends BaseClusterIntegrationTest
+    implements ExplainIntegrationTestTrait {
+
+  @BeforeClass
+  public void setUp()
+      throws Exception {
+    TestUtils.ensureDirectoriesExistAndEmpty(_tempDir, _segmentDir, _tarDir);
+
+    // Start the Pinot cluster
+    startZk();
+    startController();
+    startBroker();
+    startServers(2);
+
+    // Create and upload the schema and table config
+    Schema schema = createSchema();
+    addSchema(schema);
+    TableConfig tableConfig = createOfflineTableConfig();
+    addTableConfig(tableConfig);
+
+    // Unpack the Avro files
+    List<File> avroFiles = unpackAvroData(_tempDir);
+
+    // Create and upload segments
+    ClusterIntegrationTestUtils.buildSegmentsFromAvro(avroFiles, tableConfig, schema, 0, _segmentDir, _tarDir);
+    uploadSegments(getTableName(), _tarDir);
+
+    // Wait for all documents loaded
+    waitForAllDocsLoaded(600_000L);
+  }
+
+  protected void overrideBrokerConf(PinotConfiguration brokerConf) {
+    String property = CommonConstants.MultiStageQueryRunner.KEY_OF_MULTISTAGE_EXPLAIN_INCLUDE_SEGMENT_PLAN;
+    brokerConf.setProperty(property, "true");
+  }
+
+  @BeforeMethod
+  public void resetMultiStage() {
+    setUseMultiStageQueryEngine(true);
+  }
+
+  // Test that intermediate stages can be spooled.
+  // In this case Stage 4 is an intermediate stage whose single child is stage 5.
+  // Stage 4 is spooled and sends data to stages 3 and 7
+  @Test
+  public void intermediateSpool()
+      throws Exception {
+    JsonNode jsonNode = postQuery("SET useSpools = true;\n"
+        + "WITH group_and_sum AS (\n"
+        + "  SELECT ArrTimeBlk,\n"
+        + "    Dest,\n"
+        + "    SUM(ArrTime) AS ArrTime\n"
+        + "  FROM mytable\n"
+        + "  GROUP BY ArrTimeBlk,\n"
+        + "    Dest\n"
+        + "  limit 1000\n"
+        + "),\n"
+        + "aggregated_data AS (\n"
+        + "  SELECT\n"
+        + "    Dest,\n"
+        + "    SUM(ArrTime) AS ArrTime\n"
+        + "  FROM group_and_sum\n"
+        + "  GROUP BY\n"
+        + "    Dest\n"
+        + "),\n"
+        + "joined AS (\n"
+        + "  SELECT\n"
+        + "    s.Dest,\n"
+        + "    s.ArrTime,\n"
+        + "    (o.ArrTime) AS ArrTime2\n"
+        + "  FROM group_and_sum s\n"
+        + "  JOIN aggregated_data o\n"
+        + "  ON s.Dest = o.Dest\n"
+        + ")\n"
+        + "SELECT *\n"
+        + "FROM joined\n"
+        + "LIMIT 1");
+    JsonNode stats = jsonNode.get("stageStats");
+    Assert.assertNotNull(stats, "Stage stats should be present. Please verify the query didn't fail");
+    DocumentContext parsed = JsonPath.parse(stats.toString());
+    List<Map<String, Object>> stage4On3 = parsed.read("$..[?(@.stage == 3)]..[?(@.stage == 4)]");
+    Assert.assertNotNull(stage4On3, "Stage 4 should be a descendant of stage 3");
```

Comment:
I think the parser creates an empty list and not `null` in this case, so we can probably remove this check and rely on only the next one (`stage4On3.size() == 1`).